### ISSUE TITLE
XGBoost predictions exploding when log=Yes for tune

### ISSUE THREAD

**Issue by nicola-wiseman** [nanophyto/Abil#312]
When log=Yes, xgboost prediction for Primary Production dataset contained values 4 orders of magnitude too large, resulting in the Ens mean carrying over some of these values, leading to an integrated total that was 9 orders of magnitude too large.

ensemble_regressor.yml parameters:
 
  predictors: ["temperature",
    "sio4", "po4", "no3",
    "o2", "PAR"]

  verbose: 1
  seed : 1 # random seed
  n_threads : 10 # how many cpu threads to use, reduce when using higher cv's in MAPIE
  cv : 10
  predict_probability: False

 ensemble_config:
    classifier: False
    regressor: True
    m1: "rf"
    m2: "xgb"
    m3: "knn"

  upsample: False
  stratify: False

  reg_scoring:
   R2: r2
   MAE: neg_mean_absolute_error
   RMSE: neg_root_mean_squared_error
  predictors: ["temperature",
    "sio4", "po4", "no3",
    "o2", "PAR"]

  verbose: 1
  seed : 1 # random seed
  n_threads : 10 # how many cpu threads to use, reduce when using higher cv's in MAPIE
  cv : 10
  predict_probability: False

  param_grid:
    xgb_param_grid:
      reg_param_grid:
        n_estimators: [100]
        learning_rate: [0.1, 0.15, 0.2]
        max_depth: [5, 25, 51, 75]
        subsample: [0.4, 0.6, 0.7, 0.8, 1.0]
        colsample_bytree: [0.9, 0.95, 1]
        reg_alpha: [0, 0.1]
        gamma: [0]

**Comment by nicola-wiseman** [comment]
This did not occur previously when log=True and the hyperparameters used for that version were:
target: Primary_Production
learning_rate: 0.1
n_estimators: 100
max_depth: 51
subsample: 0.8
colsample_bytree: 0.9
gamma: 0
alpha: 0.1
transformation: 

For the version that blew up, the hyperparameters were:
target: Primary_Production
learning_rate: 0.1
n_estimators: 100
max_depth: 25
subsample: 0.4
colsample_bytree: 1
gamma: 0
alpha: 0.1
transformation: 

Also just noticed transformation parameter is not outputting for any of the 3 models. Will open as separate issue.

**Comment by nicola-wiseman** [comment]
Just realized colsample_bytree "1" should be 1.0. Will fix that...

**Comment by nanophyto** [comment]
Might be related to the objective used by XGBoost which is currently hard coded to be `objective' = reg:tweedie' `
https://github.com/nanophyto/Abil/blob/531a089bbbb217099649abc13a540cabd19a421e/abil/tune.py#L184

This means that it is using a Tweedie regression with log-link causing a double log transform:
https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters

I would suggest replacing this with a non-log transformed objective:
`reg:squarederror`

**Comment by nicola-wiseman** [comment]
> Might be related to the objective used by XGBoost which is currently hard coded to be `objective' = reg:tweedie' `
> 
> [Abil/abil/tune.py](https://github.com/nanophyto/Abil/blob/531a089bbbb217099649abc13a540cabd19a421e/abil/tune.py#L184)
> 
> Line 184 in [531a089](/nanophyto/Abil/commit/531a089bbbb217099649abc13a540cabd19a421e)
>  objective='reg:tweedie') 
> 
> This means that it is using a Tweedie regression with log-link causing a double log transform: https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters
> 
> I would suggest replacing this with a non-log transformed objective: `reg:squarederror`

Will try this.
colsample_bytree wasn't the issue because I had it was 1.0 for all my previous iterations that also had issues.

**Comment by nicola-wiseman** [comment]
Repeated error now in Predict of:
```
/opt/conda/lib/python3.11/site-packages/abil/log_grid_search.py:85: RuntimeWarning: overflow encountered in exp
  y = np.exp(x)-1
```

**Comment by nicola-wiseman** [comment]
I'm putting this done for now though to focus on CP paper writing

**Comment by ljwolf** [comment]
Overflow error again means that there are too-large values for the number representation used in numpy. This would suggest that there\should be some values that are exceptionally large (like, on the order of 1e38 in worst case, 1e308 in best case) pre-transformation. 

Is there an issue with the data? Would it be reasonable to model in terms of a different unit? e.g. if we're modeling the outcome in grams per liter, can we do that in KG/L or PG/L?

**Comment by nicola-wiseman** [comment]
Hey Levi,
I went through the data and it looks fine. It's in units of umol/m3/d, highest values being <30,000. In theory I could move to mmol/m3/d I suppose?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ljwolf** [comment]
Yes, OK—somewhere, we're then calculating `exp(30_000)`, and that's an overflow. It's very likely we're double-reversing a log transformation, calculating `exp(exp(log(y))` and overflowing. 

I will review over this week.

**Comment by nicola-wiseman** [comment]
@ljwolf Heyya, have you been able to look into this at all?

**Comment by ljwolf** [comment]
Yes. I don't see anything obviously wrong, and think the unit change is the solution. 

Would mmol/m3/d or cmol/m3/d be ok? the full range of float64 is quite accommodating at the lower end. 

We could also try casting y.astype(numpy.longdouble) if the overflow continues after casting units, but that will severely increase memory/RAM pressure.

**Comment by nanophyto** [comment]
Worth adding a warning if the training dataset has large floats?
